homeassistant/components/sensor/swiss_hydrological_data.py

```diff
@@ -143,28 +156,31 @@ def __init__(self, station):
     @Throttle(MIN_TIME_BETWEEN_UPDATES)
     def update(self):
         """Get the latest data from hydrodata.ch."""
-        from bs4 import BeautifulSoup
+        import xmltodict
 
+        details = {}
         try:
             response = requests.get(_RESOURCE, timeout=5)
         except requests.exceptions.ConnectionError:
             _LOGGER.error('Unable to retrieve data from %s', _RESOURCE)
 
         try:
-            soup = BeautifulSoup(response.content, 'html.parser')
+            stations = xmltodict.parse(response.text)['AKT_Data']['MesPar']
             # Water level: Typ="02", temperature: Typ="03", discharge: Typ="10"
-            type02, type03, type10 = [
-                soup.find(strnr='{}'.format(self.station), typ='{}'.format(i))
-                for i in ['02', '03', '10']]
-
-            details = []
-            for entry in [type02, type03, type10]:
-                details.append(entry.wert.string)
-                details.append(entry.find(typ="max24").string)
-                details.append(entry.find(typ="m24").string)
-            details.append(type03.find('name').string)
-            details.append(type03.find('zeit').string)
-
-            self.measurings = HydroData._make(details)
+            for station in stations:
+                if str(self.station) == station.get('@StrNr'):
```

Comment:
You can make this code more readable using guard clauses. Every time the if statement is False you know you are done for that item in the for-loop so you can just use `continue`:

```python
for station in stations:
    if str(self.station) != station.get('@StrNr'):
        continue

    for data in … # etc
```